### PR TITLE
Add markdown printer

### PR DESCRIPTION
**Pull request by yasunariw**
See #231.

This adds a printer that transforms the AST back into Markdown text.

Please let me know if there is a good way of adding tests to the current setup.

### CODE REVIEW

**Comment by patricoferris** [comment]
Cool feature @yasunariw ! 

I worked on something similar and you can see my findings here: https://github.com/ocaml/omd/issues/223#issuecomment-893821434. I think the conversion from `string -> Omd.doc` losses information so it is quite tricky to implement a `to_markdown` function. For example, there are some corner cases such as:

```ocaml
utop # Omd.of_string "_*hello*_" |> Omd.to_markdown;;
- : string = "**hello**\n"
```

Note this is after pinning `e60a510a54da4c3ed63da34fe6a35f6287570989`. Here, it is valid markdown to nest emphasis using the two different characters, but after using the `Omd.to_markdown` function it has been transformed into `strong` text.

What might be useful is the (very hacked together) testsuite I was running here: https://github.com/patricoferris/omd/tree/omd-print/tests. The idea was to parse the markdown normally and get the HTML, then use the `to_markdown` function and parse that output to HTML and compare the two HTML outputs. It is probably not full-proof but it was how I started finding these edge cases.

**Comment by yasunariw** [comment]
Thanks for all the pointers @patricoferris ! I'll see what else I catch using your test method and report back.

**Comment by shonfeder** [comment]
Thanks for the work on this important feature, @yasunariw and @patricoferris.

> Please let me know if there is a robust way to add tests to the current setup. Thanks!

The most robust way I can think of is via some property based testing. If we had generator that constructed valid arbitrary `Omd.doc` values, then we could check that forall `x : Omd.doc`, `(x |> Omd.to_markdown |> Omd.of_string) =  x`. This should catch the kind of edge case that @patricoferris has pointed out, iiuc.

My inclination is to suggest that we should expect any mergable solution to this feature to dependent on #223.

File: src/markdown.ml [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+  let n = infer_num_backticks 3 c in
+  nchar n buf '`';
+  let label = String.trim label in
+  Buffer.add_string buf label;
+  add_attrs_to_buffer ~space:true buf attr;
+  nl buf;
+  add_prefix ~first_item_prefix ~prefix buf;
+  add_string_with_nl_prefix ~first_item_prefix ~prefix buf c;
+  let n = String.length c in
+  if n > 0 && c.[String.length c - 1] <> '\n' then begin
```
Comment (samoht) [inline src/markdown.ml:290]:
```suggestion
  let n' = String.length c in
  if n' > 0 && c.[n'- 1] <> '\n' then begin
```

Comment (samoht) [reply src/markdown.ml:290]: otherwise the `nchar n buf` bellow adds too much backticks when closing the code block.
